_posts/2022-01-21-skyhook-bringing-computation-to-storage-with-apache-arrow.md

```diff
@@ -0,0 +1,219 @@
+Overall, the idea is simple: the client should be able to ask Ceph to perform basic operations like decoding files, filtering the data, and selecting columns.
+That way, the work gets done using existing storage cluster resources, which means it’s both adjacent to the data and can scale with the cluster size.
+Also, this reduces the data transferred over the network, and of course reduces the client workload.
+
+On the storage system side, Skyhook uses the Ceph Object Class SDK to define scan operations on data stored in Parquet or Feather format.
+To implement these operations, Skyhook first implements a file system shim in Ceph’s object storage layer, then uses the existing filtering and projection capabilities of the Arrow Datasets library on top of that shim.
+
+Then, Skyhook defines a custom “file format” in the Arrow Datasets layer.
+Queries against such files get translated to direct requests to Ceph using those new operations, bypassing the traditional POSIX file system layer.
+After decoding, filtering, and projection, Ceph directly sends the client Arrow record batches, minimizing CPU overhead for encoding/decoding—another optimization Arrow makes possible.
+The record batches use Arrow’s compression support to further save bandwidth.
+
+<figure>
+  <img src="{{ site.baseurl }}/img/20220121-skyhook-architecture.png"
+       alt="Skyhook Architecture"
+       width="100%" class="img-responsive">
+  <figcaption>
+    Skyhook extends Ceph and Arrow Datasets to push queries down to Ceph, reducing the client workload and network traffic.
+  </figcaption>
+</figure>
+
+Skyhook also optimizes how Parquet files in particular are stored.
+Parquet files consist of a series of row groups, which each contain a chunk of the rows in a file.
+When storing such files, Skyhook either pads or splits them so that each row group is stored as its own Ceph object.
+By striping or splitting the file in this way, we can parallelize scanning at sub-file granularity across the Ceph nodes for further performance improvements.
+
+## Applications
+
+In benchmarks, Skyhook has minimal storage-side CPU overhead and virtually eliminates client-side CPU usage.
+And scaling the storage cluster decreases query latency commensurately.
+For systems like Dask that use the Arrow Datasets API, this means that just by switching to the Skyhook file format, we can speed up dataset scans, reduce the amount of data that needs to be transferred, and free up CPU resources for computations.
+
+<!-- TODO: add the updated figure from Jayjeet -->
+<figure>
+  <img src="{{ site.baseurl }}/img/20220121-skyhook-architecture.png"
+       alt="In benchmarks, Skyhook reduces client CPU usage while minimally impacting storage cluster CPU usage."
+       width="100%" class="img-responsive">
+  <figcaption>
+    Skyhook frees the client CPU to do useful work, while minimally impacting the work done by the storage machines. Figure sourced from TODO.
```

Comment:
CC @JayjeetAtGithub where did this figure come from? I'd like to source it.

Reply:
@lidavidm AFAIK Carlos created it. @carlosm ?

Reply:
(I meant to put this on the figure above.) I see it in your blog post as well, so for now, I've cited that post.